### PR TITLE
WIP Add param used flag.

### PR DESCRIPTION
**Pull request by paulidale**
When parameter arrays are processed by a receiver, it is currently unknown
if the parameters could be processed or not.  This adds a used flag that
is set when a param is located.  The caller can check this to see what was
consumed.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/params.c [hunk 1]
```diff
@@ -887,3 +891,20 @@ OSSL_PARAM OSSL_PARAM_construct_end(void)
...
 }
+
+int OSSL_PARAM_used(const OSSL_PARAM *p)
+{
+    return (p->flags & OSSL_PARAM_USED) != 0;
+}
+
+void OSSL_PARAM_set_unused(OSSL_PARAM *p)
+{
+    if (p != NULL)
```
Comment (richsalz) [inline crypto/params.c]: I think if test should be removed.

File: crypto/params.c [hunk 2]
```diff
@@ -16,8 +16,10 @@ OSSL_PARAM *OSSL_PARAM_locate(OSSL_PARAM *p, const char *key)
 {
     if (p != NULL && key != NULL)
         for (; p->key != NULL; p++)
-            if (strcmp(key, p->key) == 0)
+            if (strcmp(key, p->key) == 0) {
+                p->flags |= OSSL_PARAM_USED;
```
Comment (richsalz) [inline crypto/params.c]: you need a set-used method for apps that don't call locate.  replace set-unused with this:
```
void OSSL_PARAM_set_used(OSSL_PARAM *p. int used);
```

File: doc/man3/OSSL_PARAM.pod [hunk 3]
```diff
@@ -119,6 +129,19 @@ those keys are and of what type, any functionality between those two
 entities should remain oblivious and just pass the C<OSSL_PARAM> array
 along.
 
+=head2 Flags
+
+The C<flags> field contains a number of flags:
```
Comment (richsalz) [inline doc/man3/OSSL_PARAM.pod]: "a number of" --> "the following"

File: crypto/params.c [hunk 4]
```diff
@@ -887,3 +891,20 @@ OSSL_PARAM OSSL_PARAM_construct_end(void)
...
+    return (p->flags & OSSL_PARAM_USED) != 0;
+}
+
+void OSSL_PARAM_set_unused(OSSL_PARAM *p)
+{
+    if (p != NULL)
+        p->flags &= ~OSSL_PARAM_USED;
+}
+
+void OSSL_PARAM_set_all_unused(OSSL_PARAM *params)
```
Comment (richsalz) [inline crypto/params.c]: I think this should take an "int used" flag also.

File: crypto/params.c [hunk 5]
```diff
@@ -887,3 +891,20 @@ OSSL_PARAM OSSL_PARAM_construct_end(void)
...
+{
+    while (params != NULL && params->key != NULL)
```
Comment (richsalz) [inline crypto/params.c]: I think silently ignoring a NULL parameter is wrong.

File: crypto/params.c (hunk 2 above)
Comment (paulidale) [reply crypto/params.c]: This was the alternative approach.  I'm unsure which is better.

File: doc/man3/OSSL_PARAM.pod (hunk 3 above)
Comment (paulidale) [reply doc/man3/OSSL_PARAM.pod]: Done.

File: crypto/params.c (hunk 4 above)
Comment (paulidale) [reply crypto/params.c]: A set all used defeats the purpose of this.

**Comment by paulidale** [comment]
The use case is for parameters the receiver doesn't know about.  The caller can now know if they were used or not.  #9500 raised this.

I agree that a missing critical parameter should raise an error.

File: doc/man3/OSSL_PARAM.pod [hunk 6]
```diff
@@ -119,6 +129,19 @@ those keys are and of what type, any functionality between those two
...
+=head2 Flags
+
+The C<flags> field contains the following flags:
+
+=over 4
+
+=item C<OSSL_PARAM_USED>
+
+This flag is used, by the receiver, to indicate if the parameter has been
+processed either successfully or not.
```
Comment (slontis) [inline doc/man3/OSSL_PARAM.pod]: NIT do you need the word either?

File: crypto/params.c (hunk 2 above)
Comment (paulidale) [reply crypto/params.c]: I've switch to this approach.

File: doc/man3/OSSL_PARAM.pod (hunk 6 above)
Comment (paulidale) [reply doc/man3/OSSL_PARAM.pod]: Done

**Comment by levitte** [comment]
Shouldn't functions like `OSSL_PARAM_get_int` set this flag automatically?

**Comment by paulidale** [comment]
> Shouldn't functions like OSSL_PARAM_get_int set this flag automatically?

The locate does currently.
Should the get's do as well?  I'm happy to add such.

**Comment by paulidale** [comment]
I've made the get calls set the used flag too.

File: crypto/params.c [hunk 7]
```diff
@@ -16,8 +16,10 @@ OSSL_PARAM *OSSL_PARAM_locate(OSSL_PARAM *p, const char *key)
 {
     if (p != NULL && key != NULL)
         for (; p->key != NULL; p++)
-            if (strcmp(key, p->key) == 0)
+            if (strcmp(key, p->key) == 0) {
+                OSSL_PARAM_set_used(p, 1);
```
Comment (levitte) [inline crypto/params.c]: I see no reason for this call, since you are now setting the flag in `OSSL_PARAM_get_int` etc

**Comment by paulidale** [comment]
Thinking about it, the set param calls might want to set the used flag too.  This allows the information to flow the other way and avoids problems when the returned string size is zero.

**Comment by levitte** [comment]
You do realise that OSSL_PARAM can never be const at this point, right?

**Comment by levitte** [comment]
(I.e this PR is incomplete and should probably be marked WIP)

**Comment by paulidale** [comment]
> You do realise that OSSL_PARAM can never be const at this point, right?

I thought that this assumption had died with the mutable return size change.
But yes, I did realise this.

I can see one potential use for `const OSSL_PARAM` arrays.  If we do need it, this suggested change cannot go in.  I'm fine with that but we'll still need a way to address the problems #9500 raised and how to create backward compatible ctrl functions (which do check for unknown parameters and return error when they are found).  I've also found _zero_ instances in the code thus far but will check more deeply next week -- I ran out of time today for unrelated reasons.

I need to go through the sources for `for()` loops that `strcmp` param names still, but the only build failures as a result of this were test cases that initialised the OSSL_PARAM structure directly.

**Comment by paulidale** [comment]
Argh, the get params call are passed a const OSSL_PARAM object :(  Hence the Travis failures.

**Comment by levitte** [comment]
The loss of constness means that there's no promise the receiver won't change anything in a OSSL_PARAM array. This worries me constantly.
Another solution would be to have a separate array of bits that the receiver would set for any used parameter, which would be at the caller's option to pass, or not. They would work like an FD_SET in a select() call. A consequence for this is that we need a param locator that returns an index rather than a pointer...

**Comment by paulidale** [comment]
I do not like the array of bits approach, it requires exposing more publicly.  I avoided using bit fields for similar reasons even though they would simplify the structure.  None of this is speed or space critical.  Let's make it easy for the users.

If the receiver does update other fields in the OSSL_PARAM structure they can expect a crash upon a version change.  I have zero problems with this so long as the mutability assumptions are properly documented (which they are currently).  This is a _cooperative_ arrangement between the provider and the core.

It we really do want to stop providers from changing other fields, we can make them `const`.  It will cost a few casts in the `params.c` code but I'd be quite happy to implement this.  However, I thought the idea was for the caller/owner to point to _their_ variables and a provider changing the pointers would then be ignored.

**Comment by levitte** [comment]
That last paragraph makes a good point.

**Comment by paulidale** [comment]
@levitte, which point?  That providers are cooperating or that the fields can be const.

**Comment by levitte** [comment]
Sorry, I meant last sentence

> However, I thought the idea was for the caller/owner to point to *their* variables and a provider changing the pointers would then be ignored.

**Comment by paulidale** [comment]
Okay, thanks.  I suspected that it was that one BTW.

File: crypto/params.c (hunk 7 above)
Comment (paulidale) [reply crypto/params.c]: Is it implausible for a receiver to use locate to find the param and then twiddle the fields directly?

I'm fine either way BTW.

**Comment by paulidale** [comment]
Which is preferred:

- back out the param get change or
- make param get calls not accept a const argument?

**Comment by levitte** [comment]
I don't understand the question.  This will typically be useful in calls like EVP_MD_CTX_set_params(), so the caller can check which of the parameters it sent in where actually used, yes?  That means the OSSL_PARAM needs to be writable, i.e. signatures like this:

``` C
int EVP_MD_CTX_set_params(EVP_MD_CTX *ctx, const OSSL_PARAM params[]);
```

will have to change to this:

``` C
int EVP_MD_CTX_set_params(EVP_MD_CTX *ctx, OSSL_PARAM params[]);
```

By consequense, `OSSL_PARAM_get_int()` etc will have to be changed to take a non-const OSSL_PARAM as well.

**Comment by paulidale** [comment]
That answers the question: remove the const.

**Comment by levitte** [comment]
> That answers the question: remove the const.

There was never a question, that is the consequence of this PR, there's no way around that.

**Comment by paulidale** [comment]
Okay, I see what you mean.
I've pushed a commit that removes the `const OSSL_PARAM` where necessary.

The big question now is does this adequately address the concerns in #9500 (and possible future concerns along similar lines) and is it less evil than the alternatives?

File: crypto/params.c [hunk 8]
```diff
@@ -23,11 +23,6 @@ OSSL_PARAM *OSSL_PARAM_locate(OSSL_PARAM *p, const char *key)
     return NULL;
 }
 
-const OSSL_PARAM *OSSL_PARAM_locate_const(const OSSL_PARAM *p, const char *key)
-{
-    return OSSL_PARAM_locate((OSSL_PARAM *)p, key);
-}
-
```
Comment (levitte) [inline crypto/params.c]: This is still useful, see #9303

Comment (paulidale) [reply crypto/params.c]: Okay, can put back.

Comment (paulidale) [reply crypto/params.c]: And it's back now.

**Comment by levitte** [comment]
BTW, something I just remembered is that we already have a set of functions that can be used to achieve the intent of this PR; if the providers actually implement the get_param_types / set_param_types kind of calls, the returned constant OSSL_PARAM arrays can be used to check what parameters the implementation handles. In other words, instead of flagging "I used this", the provider must declare "I will use this" in advance.

I prefer that, actually. Could we explore that first? (I'm tempted to -1 here to force that exploration)

**Comment by slontis** [comment]
> I prefer that, actually. Could we explore that first? (I'm tempted to -1 here to force that exploration)

Not quite sure I understand how this applies to individual algorithms. Setting the key len (or ivlen) for example could apply to just a small subset of the algorithms )and the ones that dont handle it can be treated generically), How does get_param_types really apply to this?

**Comment by levitte** [comment]
> Not quite sure I understand how this applies to individual algorithms. Setting the key len (or ivlen) for example could apply to just a small subset of the algorithms )and the ones that dont handle it can be treated generically), How does get_param_types really apply to this?

Just like all other functions we're dispatching, the {get,set}_param_types functions are defined per implementation (or more generic and re-used by many implementations, but that doesn't change the principal granularity).

{get,set}_param_types are supposed to return a constant OSSL_PARAM array as a description of what the {get,set}_params functions in the same implementation can handle, right?  So for example, let's take the AES-GCM vs AES-whateverelse case that was recently hit.  You can't set the key length with AES-GCM, but you can with AES-whateverelse.  So that would mean that for AES-GCM, the cipher dispatch function `OP_cipher_ctx_set_param_types` (which I realise doesn't exist yet) should return this array:

``` C
{ { OSSL_CIPHER_PARAM_AEAD_TAG, OSSL_PARAM_OCTET_STRING, NULL, 0, 0 },
  { OSSL_CIPHER_PARAM_AEAD_IVLEN, OSSL_PARAM_INTEGER, NULL, sizeof(size_t), 0 },
  { OSSL_CIPHER_PARAM_AEAD_TLS1_AAD, OSSL_PARAM_OCTET_STRING, NULL, 0, 0 },
  { OSSL_CIPHER_PARAM_AEAD_TLS1_IV_FIXED, OSSL_PARAM_OCTET_STRING, NULL, 0, 0 },
  { NULL, 0, NULL, 0, 0 } }
```

While for AES-whateverelse, it would return this:

``` C
{ { OSSL_CIPHER_PARAM_PADDING, OSSL_PARAM_INTEGER, NULL, sizeof(int), 0 },
  { OSSL_CIPHER_PARAM_NUM, OSSL_PARAM_INTEGER, NULL, sizeof(int), 0 },
  { OSSL_CIPHER_PARAM_KEYLEN, OSSL_PARAM_INTEGER, NULL, sizeof(int), 0 },
  { NULL, 0, NULL, 0, 0 } }
```

That makes it entirely visible what parameters are handled where and in what function.  My idea is that any {get,set}_params dispatched function should have a matching {get,set}_param_types (or whatever name...  basically, any dispatched function 'foo' that handles an array of OSSL_PARAM should have a matching 'foo_types' dispatched function that returns a descriptor like the above) dispatched function that returns a descriptor array, and that array would be useful for anyone who wants to check that a particular parameter they want to pass to an implementation is actually handled by said implementation.  In other words, instead of asking "was this used", we can as well ask "will this be used".

Having just looked at provider-digest(7) and provider-cipher(7), I see some serious lack of {get,set}_param_types dispatched functions.  That's unfortunate, but shouldn't be too hard to add.  provider-keymgmt(7) has a few of them, but I now see that it's also a bit incomplete.  And the providers are *seriously* lacking in this regard (side note: I have been a bit more diligent with the MAC sstuff, see #8877)

**Comment by paulidale** [comment]
The only downside to the _param_types functions is string comparisons are forced to see if a parameter is acceptable.  Not a big issue since OSSL_PARAM shouldn't be performance relevant.

I'm fine with trying an alternative before this :)

File: crypto/params.c (hunk 7 above)
Comment (paulidale) [reply crypto/params.c]: Thinking about this overnight, I concur.  The locate shouldn't set the flag.

I'm almost tempted to define the flags:

1. located
2. read
3. written

A get function sets the read flag, a set function sets the written and locate sets the located.

**Comment by paulidale** [comment]
Another overnight thought was to split the data_type field like I've done here (unsigned short, 2 x unsigned char) even if this PR goes nowhere.  Then use the other char as a OSSL_PARAM version.  This permits unknown future changes to the structure in a forward compatible manner.

I'll note that I'm not expecting any changes and can't think why we'd need or want to but better to be prepared IMHO.

I'd suggest bit fields but I'm not sure how nicely they play with ABI compatibility.

**Comment by slontis** [comment]
So if a new set param id is needed for a single algorithm, would you need to add it to every single algorithm? (This woudl not be good, especially if there are potential 3rd party providers).

**Comment by t-j-h** [comment]
@levitte you won't know what combination of things you might use until you work through it. A provider that supports a wide range of things dynamically may not be able to answer anything until such time as it is actually trying to use something.

Answering what was used is useful. Answering what *might* be used isn't really.

**Comment by levitte** [comment]
> So if a new set param id is needed for a single algorithm, would you need to add it to every single algorithm?

I assume we're talking implementations of the same algorithm here. Well, depends, right? Have we suddenly required that the ctx_set_params in every GCM implementation in the wild *must* handle the "keylen" parameter? In that perspective, is it really that bad do have another function returning a descriptor array declaring this?

@t-j-h, yes and no. We can declare what a {get,set}_params can handle at all, i.e all the keys they have any knowledge of. If a *combination* of parameters actually passed is bad, or any parameter known to the provider is bad in a given situation, then {get,set}_params should be able to detect that and signal an error. Right?

**Comment by levitte** [comment]
Basically, what this whole thing comes down to is, what parameters does a provider even look at, no matter how it's used.  The first incantation of this very PR actually wasn't so far off, with the locating functions setting the "used" flag.  That's what the keylen parameter that was mistakenly passed to the AES-GCM implementation was really all about, that there was no way to detect if that parameter is even known to that implementation.

If all parameters passed to a {get,set}_params function are known by that function (i.e. it at least looks for them), then the know that if the call is successful, the parameters were put to good use, and if the call raised an error, then we know that something was wrong, and the error reason and additional data will say the rest.  @t-j-h, doesn't this cover what you're talking about?

However, if some particular {get,set}_params doesn't have knowledge of the parameter "foo", i.e. won't even look for it, there's currently no way for the caller to know if "foo" was involved, i.e. there's uncertainty regarding the result of the call.

So I guess that it comes down to what we want to know:
1.  does a particular {get,set}_params have any knowledge of some parameter "foo", i.e. can we trust that a successful call will have looked at this parameter, or is it just pointless to pass it on?
2.  does a particular {get,set}_params not only have knowledge of some parameter "foo", but used its value?

Me, I'm hearing 1 much more loudly than 2, because 2 is kind of already answered by the outcome of a {get,set}_params call under the condition that I know what parameters it looks at; if successful, then I know that all parameters were good; if an error was raised, the error message should give me knowledge.

**Comment by paulidale** [comment]
I've pushed a version with a read, a write and a lookup flag.
Is this more in line with your thinking @t-j-h ?

**Comment by petrovr** [comment]
I use an language where is normal to send messages with property list. OpensSSL "parameter" functionality is similar.

When message is send is not clear how data is consumed. Message could send another message (nested function call), i.e. could prepare new property list based on supplied data, also it could perform property and value "translation".
When a message need data or data is not consistent it just raises error.

There is no need to ask in advance which property is required as on nested level it may depend from other properties or functionality.
With other words to check sender must have property list. And the question is why to check and the to execute as execution is enough.

Due to nested levels one property could be consumed from many levels and so on - out of scope.


In conclusion I cannot see any rationale to know when I send more then 100 X.509 certificates serialised  as OpenSSL 'parameters' which one is used in validation process.

**Comment by mattcaswell** [comment]
I am somewhat un-enthusiastic about this solution. The params structure is supposed to be really simple. It is transparent and once we release 3.0 we won't be able to make changes. The more complexity we throw in there the harder that becomes.

File: include/openssl/core.h [hunk 9]
```diff
@@ -73,17 +73,34 @@ struct ossl_algorithm_st {
...
+ */
+# define OSSL_PARAM_READ    0x0001
+# define OSSL_PARAM_WRITTEN 0x0002
+# define OSSL_PARAM_LOCATED 0x0004
+
 /* Currently supported OSSL_PARAM data types */
 /*
  * OSSL_PARAM_INTEGER and OSSL_PARAM_UNSIGNED_INTEGER
- * are arbitrary length and therefore require an arbitrarily sized buffer,
+ * are arbitrary length and therefore require an arbitrarily sized bufsfer,
```
Comment (t-j-h) [inline include/openssl/core.h]: bufsfer?

Comment (paulidale) [reply include/openssl/core.h]: Oops, no idea how that slipped in.
Fixed now.

**Comment by paulidale** [comment]
@mattcaswell any suggestion on alternative implementations that let us know what's been used?  I'm not attached to this PR but I do think we'd benefit from a solution to the underlying issue.

**Comment by mattcaswell** [comment]
> @mattcaswell any suggestion on alternative implementations that let us know what's been used? I'm not attached to this PR but I do think we'd benefit from a solution to the underlying issue.

Perhaps a simple count of the number of parameters that were processed in the return value would be sufficient.

**Comment by slontis** [comment]
> Perhaps a simple count of the number of parameters that were processed in the return value would be sufficient.

That is fine for controls which just send a single value.. Matt could you see this becoming a problem when we start passing multiple params for keys? If it processed 4 out of 5 of the params, what does that really tell us?

**Comment by paulidale** [comment]
Would it be more acceptable to revert the structure and for params.h to subvert some high order bits in the data_type field?

Pretty much everything is using params.h and we could declare the top bits to be _implementation dependent_.

Just a thought?

**Comment by mattcaswell** [comment]
> That is fine for controls which just send a single value.. Matt could you see this becoming a problem when we start passing multiple params for keys? If it processed 4 out of 5 of the params, what does that really tell us?

I see no problem with that at all. If the caller considers some params mandatory and others to not be, then just perform two calls. One with all the mandatory params, and one with all the optional params.

**Comment by richsalz** [comment]
Either things are opaque between the caller/provider, or they are not. This is be adding a "peek behind the veil" capability, which is wrong.  It's deliberately making an abstraction leaky.

If a caller wants a provider to do something, then *how it does it* should be opaque, and it should return an OK/ERROR status, and pushing a status code on the stack.

**Comment by paulidale** [comment]
> I see no problem with that at all. If the caller considers some params mandatory and others to not be, then just perform two calls. One with all the mandatory params, and one with all the optional params.

PQC algorithm barfoo takes arguments foo and bar, one is optional.  Which?
One of the points of providers is that the framework doesn't need to know.

Not that this helps here 😢

**Comment by richsalz** [comment]
> PQC algorithm barfoo takes arguments foo and bar, one is optional. Which?
> One of the points of providers is that the framework doesn't need to know.

Right, the framework doesn't need to know.  But applications have to know, there is no way around it. Saying it can be done with zero knowledge of the semantics is magical thinking.

**Comment by paulidale** [comment]
Since this seems to be unwanted, I'll close it soon if no one offers some support.

**Comment by slontis** [comment]
> I see no problem with that at all. If the caller considers some params mandatory and others to not be, then just perform two calls. One with all the mandatory params, and one with all the optional params.

@mattcaswell  - I had a look into using counters - it looks error prone. Placing count++ into every 'if locate' block inside every get_ctx_params(), set_ctx_params(), get_params() is not ideal.
Having an array of indexes would also be horrible for the caller side.
Wouldnt it just be much simpler to use param capability flags instead - similar to how it was done in the existing legacy code?

**Comment by t8m** [comment]
I really like what @levitte proposed. It could be very useful for callers to know in advance what is/can be consumed by the provider. For example various scripting frameworks could adjust their behavior accordingly. I do not think the ex-post knowledge of what was used and what not is that much useful - I mean it could be somewhat useful sometimes to verify that what I passed in was really consumed, but given it complicates the simplicity of the OSSL_PARAM I do not think it is worth it.

**Comment by paulidale** [comment]
If this were flipped around and a flag was set when the param value was changed, would it still feel dirty?

We have this capability currently (except for a zero byte octet string, the return size is set non-zero).

**Comment by levitte** [comment]
A thought I just had is that this is only a set_params (i.e. `OSSL_PARAM_get_xxx()` calls) issue, yeah?  We want to somehow know what params were actually used, yeah?  Wouldn't it be possible to do the same as we do for the other direction, simply set `return_size` to the size of the parameter data?

**Comment by slontis** [comment]
> A thought I just had is that this is only a set_params (i.e. `OSSL_PARAM_get_xxx()` calls) issue, yeah? We want to somehow know what params were actually used, yeah? Wouldn't it be possible to do the same as we do for the other direction, simply set `return_size` to the size of the parameter data?

So the gets should all return an error if the param isnt found?
I think @paulidale  has suggested something similiar in his PR?

**Comment by levitte** [comment]
> So the gets should all return an error if the param isnt found?

What?  What in what I just wrote gave you the idea I said that?

**Comment by slontis** [comment]
> What? What in what I just wrote gave you the idea I said that?

Err I think I said it :)

**Comment by slontis** [comment]
> We want to somehow know what params were actually used, yeah? Wouldn't it be possible to do the same as we do for the other direction, simply set `return_size` to the size of the parameter data?

I think that is not that much different from the pill you were having trouble swallowing :)

**Comment by levitte** [comment]
> I think that is not that much different from the pill you were having trouble swallowing :)

Oh hush, you...  :laughing: 

I think that part of that pill was how OSSL_PARAM changed an became increasingly complex.
The rest of it...  well, we do already set `return_size` in one direction, I find doing the same in the other direction a much easier pill to swallow, even though I'll mourn the disappearing `const`

**Comment by levitte** [comment]
> Was the pill blue or red?

Heh...

**Comment by richsalz** [comment]
Please don't re-use fields. Burn a flag byte.

**Comment by paulidale** [comment]
I'd prefer adding a flag but adding a field to the structure seems to be causing friction.

What I could do is rename `return_size` to `used_size` (`accessed_size`, `size` or whatever) and set it both in get and set calls.

**Comment by richsalz** [comment]
The parameter stuff is already so complicated that there are multiple/competing PR's to make it easier to create/use/maintain them. And there is friction about adding a byte? People prefer to re-use a field in hacky ways?

I hope the team reconsiders.

**Comment by paulidale** [comment]
Closing this since it isn't sufficiently wanted.
I still think the underlying problem hasn't been well addressed.

**Comment by levitte** [comment]
>  I still think the underlying problem hasn't been well addressed.

Agreed
